File: Sources/Runtime/Plugin/MelodyPlugin.swift
/// Contract for host-app plugins that register native functions into Lua under a namespace.
public protocol MelodyPlugin {
    /// The Lua namespace for this plugin (e.g., "keychain").
    /// Used as the global table name in Lua.
    var name: String { get }

    /// Called once per LuaVM to register plugin functions.
    func register(vm: LuaVM)
}

File: Android/melody-runtime/src/main/java/com/melody/runtime/plugin/MelodyPlugin.kt
package com.melody.runtime.plugin

import com.melody.runtime.engine.LuaVM

/**
 * Interface for Melody plugins that register native Lua functions.
 *
 * Each plugin exposes functions under its own top-level Lua namespace.
 * For example, a "keychain" plugin registers `keychain.get()`, `keychain.set()`, etc.
 */
interface MelodyPlugin {
    /** The Lua namespace for this plugin (e.g., "keychain"). */
    val name: String

    /** Called once per LuaVM to register plugin functions. */
    fun register(vm: LuaVM)
}
